### PR TITLE
Extract glTF utility to its own project.

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
# XRTK - Mixed Reality Toolkit Pull Request

## Overview
PR to extract the glTF utilities in to their own project.

Having a specific gLTF library within the XRTK restricted used from using their own glTF libraries and binds XRTK to tightly on a single glTF solution.

This PR simple removes the current glTF solution and moves it to it's own optional repo

## Changes

- Fixes: #800

## Breaking Changes

- Breaks
- Only breaking change is that this removes the native ability to load glTF controller models directly.  This feature should be refactored to be more open.

## Related Submodule Changes

- https://github.com/XRTK/Examples/pull/31
- https://github.com/XRTK/glTF/tree/development
- https://github.com/XRTK/WindowsMixedReality/pull/126

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs [hunk 1]
```diff
@@ -252,23 +233,10 @@ public async Task TryRenderControllerModelAsync(byte[] glbData = null, bool useA
                     ? cameraSystem.MainCameraRig.PlayspaceTransform
                     : CameraCache.Main.transform.parent;
 
-                // If the model was loaded from a system template
-                if (gltfObject != null)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: We need to provide an alternative

**Review by StephenHodgson — CHANGES_REQUESTED** [review]
See comment

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: The alternative is ultimately to use models in your project obviously.

Another way is to have in the sdk have a feature that can generate the model and then set the model parameter using a gender gltf interface.

But without a gltf system built into the core, the core can't enable that. But tbh most projects don't use this

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: I'm using the gltf to load controller models for wmr and oculus

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: So what is your suggestion to address this if you are loading models?  Given the aim is to allow users to use their preferred glTF importer.

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: Can we at least provide a game object to instantiate?

**Comment by FejZa** [comment]
My question would be, why does XRTK needs its own implementation of glTF / its own glTF module? This is something I haven't been using at all yet and thus can't really judge.

So why do we need this and what is the resolution for this PR?
@StephenHodgson  "Can we at least provide a game object to instantiate?" can you explain what you mean? We are already creating a game object for each controller aren't we?

**Comment by SimonDarksideJ** [comment]
> My question would be, why does XRTK needs its own implementation of glTF / its own glTF module? This is something I haven't been using at all yet and thus can't really judge.

The separate implementation is simply another variant on the Khronos implementation but updated to use Unity JSON and a host of other improvements.  I think it's worth preserving and maintaining in its own repo.  But being separate means you can use your own if you wish instead.

> So why do we need this and what is the resolution for this PR?
> @StephenHodgson "Can we at least provide a game object to instantiate?" can you explain what you mean? We are already creating a game object for each controller aren't we?

That's what confused me, just hadn't had the time to circle back to this conversation.  Any further thoughts @StephenHodgson ?

**Comment by StephenHodgson** [comment]
Yes this is true, for edit time, but if we have a 3rd party glTF importer it'll likely pass a GameObject reference as either the instantiated object or an object to insatiate. I suppose that's what I'm looking for exactly. Just to link the references up (if the object has already been created) or to create it for us if it's not already in the scene.

**Comment by FejZa** [comment]
Fair enough, now I get it.

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: OK, doing some testing and the simplest way would be for the developer to update the visualizer type for the controller to be one to load the glTF model required.

You don't need a GameObject to attach to, simply use your own visualizer that loads your glTF model.

Thoughts @StephenHodgson

**Comment by SimonDarksideJ** [comment]
> but if we have a 3rd party glTF importer it'll likely pass a GameObject reference

@StephenHodgson  As commented above, you don't need a GameObject reference, just write your own Visualizer to load the glTF model and set the Visualizer type to your new visualiser for the controllers you want to configure.

File: XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs (hunk 1 above)
Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: Where is `gltfObject` declared? Shouldn't we remove that as well?

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Runtime/Providers/Controllers/BaseController.cs:256]: > Where is `gltfObject` declared? Shouldn't we remove that as well?

Line 222, it was removed.  THere is no longer any glTF reference in the Core classes post this change

**Comment by StephenHodgson** [comment]
What is failing the build?

File: XRTK-Core/Packages/com.xrtk.core/Editor/BuildAndDeploy/BuildDeployWindow.cs [hunk 2]
```diff
@@ -525,16 +524,16 @@ private void AppxBuildGUI()
 
             var prevFieldWidth = EditorGUIUtility.fieldWidth;
 
-            EditorGUIUtility.fieldWidth = 120;
-            var appIconAssetObject = AssetDatabase.LoadAssetAtPath<GltfAsset>(UwpBuildDeployPreferences.MixedRealityAppIconPath);
+            //EditorGUIUtility.fieldWidth = 120;
+            //var appIconAssetObject = AssetDatabase.LoadAssetAtPath<GltfAsset>(UwpBuildDeployPreferences.MixedRealityAppIconPath);
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Editor/BuildAndDeploy/BuildDeployWindow.cs:529]: Let's just go ahead and remove the lines altogether. I'm rewriting this anyway and don't want conflicts.

Comment (SimonDarksideJ) [reply XRTK-Core/Packages/com.xrtk.core/Editor/BuildAndDeploy/BuildDeployWindow.cs:529]: Darn, I thought I had deleted everything.  Good spot.

**Comment by SimonDarksideJ** [comment]
> What is failing the build?

Just checked, it's complaining about the Examples project.  I'll do a quick PR to also remove the glTF samples from there

**Comment by SimonDarksideJ** [comment]
Pushed an update to change the branch for examples and the recent comments

**Comment by SimonDarksideJ** [comment]
@StephenHodgson Issues resolved

**Review by StephenHodgson — CHANGES_REQUESTED** [review]
Let's keep the glTF examples and just update the references

**Comment by StephenHodgson** [comment]
@SimonDarksideJ let's be sure to create a issue in WMR for adding the visualizer later on once the glTF package is published.

**Review by StephenHodgson — APPROVED** [review]

**Pull request merged**